### PR TITLE
Feature Multi state buttons

### PR DESCRIPTION
**Pull request by rkmax**
## Description of the changes

**⚠️ Important note**: this PR introduce breaking changes to the configuration file, the changes are not backwards compatible, the apllication will take care of migrating your file of version `1`. Be aware of backup or export you current configuration before trying this changes

API Changes

- Add support of multi state buttons
- Add switch state action to button, allowing to change between states
- Button state creating and removal works in the same way that pages
- In the same way that the default font is not stored I replicate this behavior for  font color, font size and background color

GUI Changes

- A tab section that allow switch between button states, works like pages
- Button that allow remove/add button states
- Switch state input, allow to change to that button state
- now when type on any inputs like (Label, Command, Press Keys, Write text) the app will be more responsibe

CLI changes

- add  SET_STATE to the option --action 
- add --state option
- Allow change button state using CLI

## Motivation of the factor of API state

changes needed in order to add the feature that is adding this PR are not trivial, to implement them the state should be managed correctly, i decide to use dataclasses which are a feature of python to handling data, they are easy to serialize/deserialize, now we dont have to silent linter and autocomplete functionality should work better for anyone interested in contribute

## Notes for reviewers

- I added tests for the functionalities of the API (device state, button state, config) I split the tests grouped by features with the goal to not create endless simple file of tests
- I tried to commit the changes the more cohesive possible. I know that some of the changes are still splited but hopefully they are clear enough to be reviewed without too much hassle
- in the refactor I renamed some of the arguments used across the API which give me some confusion when I started to get hands in the code, like the deck_id and serial_number, deck_id is basically just used for dettaching events the rest of the app uses serial_number, this naming change occours only at API level, I did some improvements to the GUI in general to make more constent, like namings, validations checks, centralize logic in a future improvements we can refactor in clasess or split the code into multiple files

# Demo

image

https://github.com/streamdeck-linux-gui/streamdeck-linux-gui/assets/395707/1b2c33c8-021d-4757-912e-5ed1059c9922

### CODE REVIEW

**Comment by rkmax** [comment]
I notice the select installed font from system was merged. I'll need to solve these conflicts here since I refactored gui.py. I'll let you know when is done for review

**Comment by Quintar** [comment]
Just wondering, is it possible to switch states from the outside? (Like a weather script that changes the state to the "sunny" or "rainy" state of the button)

**Comment by rkmax** [comment]
> Just wondering, is it possible to switch states from the outside? (Like a weather script that changes the state to the "sunny" or "rainy" state of the button)

thank you for remember me this, I forgot to update the cli. command `streamdeckc` which will allow you switch properties

**Review by 4s3ti — CHANGES_REQUESTED** [review]
This one going to have quite some text :p 


Is shaping up nicely. I am not commenting on the coding itself because its quite big and obviously goes beyond my knowledge of python and my understanding of the whole codebase, however I can notice a clear concern into making it cleaner and easier to understand. really appreciated. 

did some testing and works quite nice, just found something ...  

Remove Image button is borked, when pressed does nothing and spits the following error: 

```TypeError: show_button_state_remove_image_dialog() takes 0 positional arguments but 1 was given```


Being this such a big Pull request with a breaking change this will actually become a major release and will make this V4. Which I am happy with :)   

My suggestion is that when you are done, you actually squash all the commits and submit it as a single big feature, the test commits can be included on it. 

The commit should have in the body message "BREAKING CHANGE: <Description of the change>" 

Feel free to include a HUGE body in the commit with all the feature introduces, I am okay with that. 

I will hold this PR for a bit (and all others) as I want to start creating beta releases instead of the develop leaving develop branch as an actual development for actual heavy development that takes in all the stuff. 

It doesn't look that the migration process is making any backup or displaying any warning, therefore I suggest that before touching the configuration file we should protect the users from shooting themselves on the foot. 

* When older version of the config file is detected display a warning to the user and allow it to terminate the app (so the user can create a backup by himself)
* If the user accepts the changes, before migrating, we create a backup ourselves `streamdeck_ui.json_old`


Also please state if there are any notable changes on the dependencies that might affect packages such as AUR

feel free to ping me directly on discord if you have any questions better suited for a chat session or you need any quick answer.

**Comment by rkmax** [comment]
> This one going to have quite some text :p
> 
> Is shaping up nicely. I am not commenting on the coding itself because its quite big and obviously goes beyond my knowledge of python and my understanding of the whole codebase, however I can notice a clear concern into making it cleaner and easier to understand. really appreciated.
> 
> did some testing and works quite nice, just found something ...
> 
> Remove Image button is borked, when pressed does nothing and spits the following error:
> 
> `TypeError: show_button_state_remove_image_dialog() takes 0 positional arguments but 1 was given`
> 
> Being this such a big Pull request with a breaking change this will actually become a major release and will make this V4. Which I am happy with :)
> 
> My suggestion is that when you are done, you actually squash all the commits and submit it as a single big feature, the test commits can be included on it.
> 
> The commit should have in the body message "BREAKING CHANGE: "
> 
> Feel free to include a HUGE body in the commit with all the feature introduces, I am okay with that.
> 
> I will hold this PR for a bit (and all others) as I want to start creating beta releases instead of the develop leaving develop branch as an actual development for actual heavy development that takes in all the stuff.
> 
> It doesn't look that the migration process is making any backup or displaying any warning, therefore I suggest that before touching the configuration file we should protect the users from shooting themselves on the foot.
> 
> * When older version of the config file is detected display a warning to the user and allow it to terminate the app (so the user can create a backup by himself)
> * If the user accepts the changes, before migrating, we create a backup ourselves `streamdeck_ui.json_old`
> 
> Also please state if there are any notable changes on the dependencies that might affect packages such as AUR
> 
> feel free to ping me directly on discord if you have any questions better suited for a chat session or you need any quick answer.

thank you so much for the review. I'll take in consideration all the points that you mentioned. I agree regarding the desition of allow the user to choose if migrate the config or not. as well like in case said yes to backup the old file

**Comment by 4s3ti** [comment]
fyi, Development branch is now ready, and the release process should be a bit more efficient :D

**Comment by ajbucci** [comment]
@rkmax Font preview in font selection combobox didn't make it into the merge. Fix is simply replacing:

`ui.text_font.setItemData(i, font)`
with
`ui.text_font.setItemData(i, font, Qt.FontRole)  # type: ignore [attr-defined]`
in gui.py

File: streamdeck_ui/gui.py
Comment (ajbucci) [inline streamdeck_ui/gui.py:643]: ui.text_font.setItemData(i, font, Qt.FontRole)  # type: ignore [attr-defined]

File: streamdeck_ui/gui.py
Comment (ajbucci) [inline streamdeck_ui/gui.py:1191]: ui.text_font.setItemData(i, font, Qt.FontRole)  # type: ignore [attr-defined]

Comment (Data-007) [reply streamdeck_ui/gui.py:1191]: Config file blank, try to add new page broken
#82-BrokenPageAdd

**Comment by rkmax** [comment]
@ajbucci thank you for your feedback. I'll take care of that

File: streamdeck_ui/gui.py
Comment (rkmax) [reply streamdeck_ui/gui.py]: it should be fixed now. please if you can confirm that it's fixed just mark as resolve the conversation. thank you in advance

File: streamdeck_ui/gui.py
Comment (rkmax) [reply streamdeck_ui/gui.py:647]: it should be fixed now. please if you can confirm that it's fixed just mark as resolve the conversation. thank you in advance

File: streamdeck_ui/gui.py
Comment (Data-007) [reply streamdeck_ui/gui.py]: Yes All fixed now thanks

**Comment by Data-007** [comment]
Alls good seems to work, no pop to guide me to backup file but iam on V2 anyway

File: streamdeck_ui/gui.py
Comment (Data-007) [reply streamdeck_ui/gui.py]: Yup fixed thanks

**Comment by 4s3ti** [comment]
The CI Tool that parses the commit messages to generate release notes and automatically version the code most likely probably won't pick up the `BREAKING CHANGE` on the commit message body. 

as for feat/fix/build etc ... only matters on the Commit message header. 

If I understand correctly the move from dicts to data classes is the actual breaking change as it affects how the JSON is generated right? 

My suggestion is: 

```
BREAKING CHANGE: Use dataclasses instead of dicts to handle configuration and support multi state buttons.


gui: configuration now support multiple button states, v1 state file will be migrated, v2 is not backward compatible
gui: warn dialog about migrating file and backup
gui: allow add and remove button states
gui: allow set the action "switch state" in button settings
cli: allow set the button state


bulid
add ubuntu libraries for virtual framebuffer
add python pytest-qt libraries

refactor

fix
minor bugs resulting in bad buttons display under some circumstances

test
api: add test for add and remove button states
api: add test for add and remove pages
api: add test for change all properties (font, font size, color, etc)
config: test read/write and migration configuration
gui: add test for all menu items
gui: add tests for pages actions (add and remove)
gui: add tests for button states actions add and remove
gui: add tests for warn migration message
```

**Comment by rkmax** [comment]
> The CI Tool that parses the commit messages to generate release notes and automatically version the code most likely probably won't pick up the `BREAKING CHANGE` on the commit message body.
> 
> as for feat/fix/build etc ... only matters on the Commit message header, which yours is perfect, don't change it!
> 
> If I understand correctly the move from dicts to data classes is the actual breaking change as it affects how the JSON is generated right?
> 
> My suggestion is:
> 
> ```
> feat: Multi state buttons
> 
> BREAKING CHANGE: Use dataclasses instead of dicts to handle configuration and support multi state buttons.
> 
> config: state now uses models instead dicts
> api: state now uses models instead dicts
> gui: configuration now support multiple button states, v1 state file will be migrated, v2 is not backward compatible
> gui: warn dialog about migrating file and backup
> gui: allow add and remove button states
> gui: allow set the action "switch state" in button settings
> gui: cleaner logic to update button states
> cli: allow set the button state
> build/ci: add ubuntu libraries for virtual framebuffer
> build/ci: add python pytest-qt libraries
> api: add test for add and remove button states
> api: add test for add and remove pages
> api: add test for change all properties (font, font size, color, etc)
> gui: add test for all menu items
> gui: add tests for pages actions (add and remove)
> gui: add tests for button states actions add and remove
> gui: add tests for warn migration message
> config: test read/write and migration configuration
> fix minor bugs resulting in bad buttons display under some circumstances
> ```

thank you for the reference commit. I just updated. lmk if everything it's correct

**Review by 4s3ti — APPROVED** [review]

**Comment by 4s3ti** [comment]
Thank you a lot for your work! :)

**Pull request merged**
